src/dimension.c

```diff
@@ -1646,14 +1711,145 @@ ts_dimension_add(PG_FUNCTION_ARGS)
 		}
 	}
 
-	ts_hypertable_func_call_on_data_nodes(info.ht, fcinfo);
+	ts_hypertable_func_call_on_data_nodes(info->ht, fcinfo);
 
-	retval = dimension_create_datum(fcinfo, &info);
+	retval = dimension_create_datum(fcinfo, info, is_generic);
 	ts_cache_release(hcache);
 
 	PG_RETURN_DATUM(retval);
 }
 
+TS_FUNCTION_INFO_V1(ts_dimension_add);
+TS_FUNCTION_INFO_V1(ts_dimension_add_general);
+
+Datum
+ts_dimension_add(PG_FUNCTION_ARGS)
+{
+	DimensionInfo info = {
+		.type = PG_ARGISNULL(2) ? DIMENSION_TYPE_OPEN : DIMENSION_TYPE_CLOSED,
+		.table_relid = PG_GETARG_OID(0),
+		.num_slices = PG_ARGISNULL(2) ? DatumGetInt32(-1) : PG_GETARG_INT32(2),
+		.num_slices_is_set = !PG_ARGISNULL(2),
+		.interval_datum = PG_ARGISNULL(3) ? Int32GetDatum(-1) : PG_GETARG_DATUM(3),
+		.interval_type = PG_ARGISNULL(3) ? InvalidOid : get_fn_expr_argtype(fcinfo->flinfo, 3),
+		.partitioning_func = PG_ARGISNULL(4) ? InvalidOid : PG_GETARG_OID(4),
+		.if_not_exists = PG_ARGISNULL(5) ? false : PG_GETARG_BOOL(5),
+	};
+
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+
+	if (!PG_ARGISNULL(1))
+		memcpy(&info.colname, PG_GETARG_NAME(1), NAMEDATALEN);
+
+	if (PG_ARGISNULL(0))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("hypertable cannot be NULL")));
+
+	return ts_dimension_add_internal(fcinfo, &info, false);
+}
+
+TSDLLEXPORT Datum
+ts_dimension_info_in(PG_FUNCTION_ARGS)
+{
+	Oid argtype = get_fn_expr_argtype(fcinfo->flinfo, 0);
+	ereport(ERROR,
+			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+			 errmsg("cannot accept a value of type OID \"%d\"", argtype)));
+
+	PG_RETURN_VOID(); /* keep compiler quiet */
+}
+
+TSDLLEXPORT Datum
+ts_dimension_info_out(PG_FUNCTION_ARGS)
+{
+	DimensionInfo *info = (DimensionInfo *) PG_GETARG_POINTER(0);
+	StringInfoData str;
+
+	switch (info->type)
+	{
+		case DIMENSION_TYPE_CLOSED:
+			appendStringInfo(&str,
+							 "hash//%s//%d//%s",
+							 NameStr(info->colname),
+							 info->num_slices,
+							 get_func_name(info->partitioning_func));
+			break;
+
+		case DIMENSION_TYPE_OPEN:
+		{
+			bool isvarlena;
+			Oid outfuncid;
+
+			getTypeOutputInfo(info->interval_type, &outfuncid, &isvarlena);
+			Assert(OidIsValid(outfuncid));
+			const char *argvalstr = OidOutputFunctionCall(outfuncid, info->interval_datum);
+
+			appendStringInfo(&str,
+							 "range//%s//%s//%s",
+							 NameStr(info->colname),
+							 argvalstr,
+							 get_func_name(info->partitioning_func));
+			break;
+		}
+
+		case DIMENSION_TYPE_ANY:
+			appendStringInfo(&str, "any");
+			break;
+	}
+	PG_RETURN_CSTRING(str.data);
+}
+
+/*
+ * DimensionInfo for a hash dimension.
+ *
+ * This structure is only partially filled in when constructed. The rest will
+ * be filled in by ts_dimension_add_general.
+ */
+Datum
+ts_hash_dimension(PG_FUNCTION_ARGS)
+{
+	Name colname = PG_GETARG_NAME(0);
+	DimensionInfo *info = palloc0(sizeof(DimensionInfo));
+	Ensure(PG_NARGS() > 2, "expected 3 arguments, defined with %d", PG_NARGS());
```

Comment:
I am not sure if `defined with %d` is the right term here. Maybe something like `invoked with %d arguments` is better to understand?

Nit: I would suggest performing the `Ensure` before `palloc0` is called and failing before we have asked for memory.

Reply:
Fixed.